### DISCUSSION TITLE
How does R-universe determine system dependencies?

### DISCUSSION THREAD

**Discussion by rdinnager** [r-universe-org/help#99] [Q&A] [answered]
Hi all,

I'm really liking having R-universe as an alternative to CRAN for distributing my packages, but I ran into a problem with one of mine: https://github.com/r-universe/rdinnager/actions/runs/1327222576

It failed to build because a system dependency wasn't installed. This is possibly because I didn't specify it correctly in the SystemRequirements: entry of DESCRIPTION (I note that system dependencies were installed correctly for the R package dependencies of my package, e.g. sf, imager, etc.). Note that the system dependency I need is somewhat non-standard. In fact, I added it myself to rtools-packages (https://github.com/r-windows/rtools-packages/pull/223) in order to make my package easier to install on Windows. The required package is also available on MacOS via brew. 

So I guess my question is: "How can I make sure that R-universe knows what system dependencies to install, and where to get them?"

Apologies if there is a document out there already with information on this that I have missed, and thanks for all your work on this.

**Comment by drmowinckels** [comment]
I'd really like some document about this too. 
I currently have a package whose source build is failing with vignettes because orca command line tool is not installed, understandably. 
But it is fairly easy to amend, if instructions on how to do that were supplied.

https://github.com/r-universe/ggseg/runs/3909538684?check_suite_focus=true#step:3:1583

**Comment by jeroen** [comment, marked as answer]
Our build environment is very similar to the CRAN environment. For Linux that means that most system libraries that are needed by R packages are preinstalled on the build image. We also try to query rstudio's [r-system-requirements](https://github.com/rstudio/r-system-requirements) on linux for any additional libraries.

On Windows and MacOS our build image is also very similar to CRAN. A few libraries are preinstalled, but most packages that need additional libraries will download a suitable version at install time. I have implemented this for @rdinnager rbbf in https://github.com/rdinnager/rbff/pull/3.

The reason why we don't provide some hook to let package authors install arbitrary things on our build server (as you may do in your own CI) is that we want the R package that are published on r-universe, to also work the same way on user's machines, and not depend on some external component or library that most users may not have. Therefore we do not support e.g. homebrew.

@Athanasiamo if orca is only needed to render the vignette, have you considered pre-computing the vignette as explained on e.g. https://ropensci.org/blog/2019/12/08/precompute-vignettes/ ? This way you could get the package published also on CRAN (I'm quite sure they also don't have orca).

**Reply by drmowinckels** [reply]
Thanks for the reply. This makes sense. Orca is not just for vignette, but I didnt know about prebuilding vignettes, so thank you for the link. i'll have a look.

Also, thank you for the rstudio r-system-requirements link, that is super interesting and I had no idea about it. These two link might just solve the immediate needs I have.

**Reply by drmowinckels** [reply]
fyi, this particular package I'm on about will likely never be cran viable, which is another reason r-universe is so dope!
